src/main/java/com/walmartlabs/x12/standard/AbstractStandardX12Parser.java

```diff
@@ -0,0 +1,308 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.StringUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * X12 Parser
+ *
+ * Envelope
+ * -- ISA
+ * ----- Groups
+ * -- ISE
+ *
+ * Group
+ * -- GS
+ * ----- Transactions
+ * -- GE
+ *
+ * Transaction
+ * -- ST
+ * ----- Transaction Details
+ * -- SE
+ *
+ */
+public abstract class AbstractStandardX12Parser<T extends StandardX12Document> implements X12Parser<T> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractStandardX12Parser.class);
+
+    public static final String ISA_HEADER_ID = "ISA";
+    public static final String ISA_TRAILER_ID = "IEA";
+
+    public static final String GROUP_HEADER_ID = "GS";
+    public static final String GROUP_TRAILER_ID = "GE";
+
+    public static final String TRANSACTION_HEADER_ID = "ST";
+    public static final String TRANSACTION_TRAILER_ID = "SE";
+
+    /**
+     * parse an X12 document into the representative Java POJO
+     *
+     * @return T (the Class associated with the parser)
+     * @throws X12ParserException if the document can't be parsed
+     */
+    @Override
+    public T parse(String sourceData) {
+        StandardX12Document x12Doc = null;
+
+        try {
+            if (!StringUtils.isEmpty(sourceData)) {
+                // delegate creation of the concrete X12 document
+                x12Doc = this.createX12Document();
+
+                // break document up into segment lines
+                SegmentUtil segments = new SegmentUtil(this.splitSourceDataIntoSegments(sourceData));
+
+                // standard parsing of segment lines
+                this.standardParsingTemplate(segments, x12Doc);
+            }
+        } catch (Exception e) {
+            throw new X12ParserException("Invalid EDI X12 message: unexpected error", e);
+        }
+
+        return (T) x12Doc;
+    }
+
+    /**
+     * template for parsing a standard EDI X12 document
+     *
+     * @throws X12ParserException if the document can't be parsed
+     */
+    private void standardParsingTemplate(SegmentUtil segments, StandardX12Document x12Doc) {
+
+        //
+        // interchange control header
+        //
+//        int segmentIdx = 0;
+//        X12Segment currentSegment = this.nextSegment(segmentIdx++, segmentLines);
+        X12Segment currentSegment = segments.nextSegment();
+        this.parseInterchangeControlHeader(currentSegment, x12Doc);
+
+        //
+        // parse groups
+        //
+        boolean insideGroup = false;
+        boolean insideTransaction = false;
+
+        while (segments.hasMoreSegmentLines()) {
+            // get the next segment
+//            currentSegment = this.nextSegment(segmentIdx++, segmentLines);
+            currentSegment = segments.nextSegment();
+
+            // parse group header
+            X12Group currentGroup = this.parseGroupHeader(currentSegment, x12Doc);
+            x12Doc.addGroupHeader(currentGroup);
```

Comment:
Do you(@CodeSmell)  think that `x12Doc.addGroupHeader(currentGroup);` should be later when we are done parsing group or it does not matter?

Reply:
Don't think it should matter